bot/utils/scheduling.py

```diff
@@ -1,81 +1,112 @@
 
-from bot.utils import CogABCMeta
 
-log = logging.getLogger(__name__)
+class Scheduler:
+    """
+    Schedule the execution of coroutines and keep track of them.
 
+    When instantiating a Scheduler, a name must be provided. This name is used to distinguish the
+    instance's log messages from other instances. Using the name of the class or module containing
+    the instance is suggested.
 
-class Scheduler(metaclass=CogABCMeta):
-    """Task scheduler."""
+    Coroutines can be scheduled immediately with `schedule` or in the future with `schedule_at`
+    or `schedule_later`. A unique ID is required to be given in order to keep track of the
+    resulting Tasks. Any scheduled task can be cancelled prematurely using `cancel` by providing
+    the same ID used to schedule it.  The `in` operator is supported for checking if a task with a
+    given ID is currently scheduled.
 
-    def __init__(self):
-        # Keep track of the child cog's name so the logs are clear.
-        self.cog_name = self.__class__.__name__
+    Any exception raised in a scheduled task is logged when the task is done.
+    """
 
-        self._scheduled_tasks: t.Dict[t.Hashable, asyncio.Task] = {}
-
-    @abstractmethod
-    async def _scheduled_task(self, task_object: t.Any) -> None:
-        """
-        A coroutine which handles the scheduling.
+    def __init__(self, name: str):
+        self.name = name
 
-        This is added to the scheduled tasks, and should wait the task duration, execute the desired
-        code, then clean up the task.
+        self._log = logging.getLogger(f"{__name__}.{name}")
+        self._scheduled_tasks: t.Dict[t.Hashable, asyncio.Task] = {}
 
-        For example, in Reminders this will wait for the reminder duration, send the reminder,
-        then make a site API request to delete the reminder from the database.
-        """
+    def __contains__(self, task_id: t.Hashable) -> bool:
+        """Return True if a task with the given `task_id` is currently scheduled."""
+        return task_id in self._scheduled_tasks
 
-    def schedule_task(self, task_id: t.Hashable, task_data: t.Any) -> None:
+    def schedule(self, task_id: t.Hashable, coroutine: t.Coroutine) -> None:
         """
-        Schedules a task.
+        Schedule the execution of a coroutine.
 
-        `task_data` is passed to the `Scheduler._scheduled_task()` coroutine.
+        If a task with `task_id` already exists, close `coroutine` instead of scheduling it.
+        This prevents unawaited coroutine warnings.
         """
-        log.trace(f"{self.cog_name}: scheduling task #{task_id}...")
+        self._log.trace(f"Scheduling task #{task_id}...")
 
         if task_id in self._scheduled_tasks:
-            log.debug(
-                f"{self.cog_name}: did not schedule task #{task_id}; task was already scheduled."
-            )
+            self._log.debug(f"Did not schedule task #{task_id}; task was already scheduled.")
+            coroutine.close()
```

Comment:
This ties into this https://github.com/python-discord/bot/issues/800#issuecomment-647047778. After writing that comment, I recognised these other problems with the current approach but felt like it was too much of an edge case to care. Maybe not.

I don't want to make it private in case someone wants to schedule a coroutine immediately and still take advantage of the scheduler keeping track of it.

Reply:
> I don't want to make it private in case someone wants to schedule a coroutine immediately and still take advantage of the scheduler keeping track of it.

That's definitely fair enough, but in that case, you probably want to at least leave a block comment above the `coroutine.close()`. I could see it being rather difficult to debug if someone mistakenly passes a coroutine that has started running or might be awaited later.

And maybe an `assert inspect.getcoroutinestate(coroutine) == "CORO_CREATED"` before it, which would be removed when the code is ran with the `-O` flag. That could help to ensure safe usage in testing, while removing any production runtime impact if you're certain that in most cases it would be a recently created coroutine.